[Bug 18335] Give Xcode setup script ability to cache SDKs

peter-b opened the pull request:
Xcode app bundles are pretty big, and it's a pain to have to keep half a dozen of them around in order to build LiveCode.  This PR modifies the Xcode SDK setup script so that if you pass it the `--cache` argument it:
1. copies all of the SDKs it needs into a cache directory
2. symlinks the cached SDKs into the target Xcode app bundle.

If you don't run it in `--cache` mode then its behaviour is the same as before (apart from being written in Python rather than shell).

montegoulding commented on tools/setup_xcode_sdks.py:
```diff
@@ -0,0 +1,200 @@
+
+# Update these lists if you need different SDK versions!
+
+iphoneos_versions = ["10.0", "9.3", "9.2", "8.2"]
```
@peter-b I'm wondering if this can be implemented in a way that just looks for SDKs rather than requiring maintenance when there's a new SDK?

peter-b replied:
@montegoulding IMHO it's important that this script fails if it can't find all the SDKs needed for a successful build.  I think this script is the only place where that info is stored (I think the iOS SDK versions could maybe be parsed from the top-level Makefile, but I'm pretty certain that the MacOS SDK versions aren't available anywhere but here).

Do we need to deal with this as part of this PR?

montegoulding replied:
@peter-b ok and no

montegoulding commented on tools/setup_xcode_sdks.py:
```diff
@@ -0,0 +1,200 @@
+# 5) Run this script:
+#
+#        cd ${dir}
+#        sh /path/to/setup_xcode_sdks.sh
```
`.py` ?

peter-b replied:
The shell script now redirects to the Python program.

montegoulding commented on tools/setup_xcode_sdks.sh:
```diff
@@ -1,118 +1,3 @@
-  echo "ERROR: Some SDKs couldn't be installed" >&2
-  exit 1
-fi
+exec python ${0%.sh}.py $@
```
Ah... never mind the `.py` comment

montegoulding commented:
@livecode-vulcan review ok 6074b29

livecode-vulcan commented:
💙 review by @montegoulding ok 6074b294448099e7bcecef18e925280b391f533f

livecode-vulcan commented:
😞 test failure 6074b294448099e7bcecef18e925280b391f533f
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/170)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/174)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/170)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/170)
- try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/174)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/170)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/170)

montegoulding commented:
I guess this needs to be rebased onto the current develop-8.0 head

peter-b committed:
[Bug 18335] Convert Xcode SDK setup script to Python

This corrects a bunch of problems when working with Xcode app
bundles stored in directories containing spaces...

peter-b committed:
[Bug 18335] Add SDK caching to setup_xcode_sdks.py tool

If you pass `--cache` as an argument to the Xcode SDK setup script,
it will now copy all referenced SDKs into a cache directory before
symlinking them into the target Xcode app bundle.

peter-b commented:
@montegoulding I've rebased this and targeted it at develop-8.1.  Could you please re-review?

montegoulding commented:
@livecode-vulcan review ok c9442bd

livecode-vulcan commented:
💙 review by @montegoulding ok c9442bd6769397c7880ea282b9a800e6fe93d051

livecode-vulcan commented:
😎 test success c9442bd6769397c7880ea282b9a800e6fe93d051
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/190)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/194)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/190)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/190)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/194)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/190)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/190)
